walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,234 @@
+    // Step 1: Find all occupied slots of relevent people based on the list of attendees
+    ArrayList<TimeRange> allOccupiedSlots = getAllOccupiedSlots(commonCalendar, attendees);
+
+    // Step 2: Find all possible windows by merging occupied ones and fine the (inverse selection) of those times
+    // Step 2a: Sort the allOccupiedSlots
+    Collections.sort(allOccupiedSlots, TimeRange.ORDER_BY_START);
+
+    // Step 2b: Get a concise version of all occupied slots (merge overlapping ones)
+    allOccupiedSlots = getOverlappedSlots(allOccupiedSlots);
+
+    // Step 2c: Get the available slots
+    ArrayList<TimeRange> allAvailableSlots = getInverseSlots(allOccupiedSlots);
+    return allAvailableSlots;
   }
+
+  /**
+  * @return an arrayList of all occupied slots from a common calendar of attendees
+  */
+  private ArrayList<TimeRange> getAllOccupiedSlots(HashMap<String, ArrayList<TimeRange>> calendar, Collection<String> attendees){
+    ArrayList<TimeRange> allOccupiedSlots = new ArrayList<TimeRange>();
+    for (String attendee: attendees){
+        if (calendar.containsKey(attendee)) {
+            ArrayList<TimeRange> occupiedSlots = calendar.get(attendee);
+            for (int i = 0; i < occupiedSlots.size(); i++) {
+                allOccupiedSlots.add(occupiedSlots.get(i));
+            }
+        } else {// The attendee is not in the database yet, which means this attendee is assumed to be free all day
+            allOccupiedSlots.add(TimeRange.fromStartEnd(0, 0, false));
+        } 
+    }
+    return allOccupiedSlots;
+  }
+  
+  /**
+  * @param an ArrayList of timeslots sorted by start, can be occupied or unoccupied
+  * @return all overlap time slots 
+  */
+  private ArrayList<TimeRange> getOverlappedSlots(ArrayList<TimeRange> timeSlots) {
+      ArrayList<TimeRange> overlappedSlots = new ArrayList<TimeRange>();
+      if (timeSlots.isEmpty()){
+          return overlappedSlots;
+      }
+
+      // Start from the first time slot, return the most concise treeset of time slots (that are either occupied or not)
+      TimeRange overlappedSlot = timeSlots.get(0);
+
+      if (timeSlots.size() == 1) {
+          overlappedSlots.add(overlappedSlot);
+          return overlappedSlots;
+      }
+      
+      boolean inclusive = false;
+      // There are at least two things in timeSlots
+      for (int i = 1; i < timeSlots.size(); i++){
+
+          if (timeSlots.get(i).overlaps(overlappedSlot)){
+              if (timeSlots.get(i).end() == TimeRange.END_OF_DAY) {inclusive = true;}
+              overlappedSlot = overlappedSlot.mergeOverlapped(timeSlots.get(i), inclusive);
+              if (i == timeSlots.size() - 1) {// If this is the last time slot
+                    overlappedSlots.add(overlappedSlot);
+              }
+          } else {
+              overlappedSlots.add(overlappedSlot);
+              overlappedSlot = timeSlots.get(i);
+              if (i == timeSlots.size() - 1) {
```

Comment:
this is a common logic in both if and else, so you can move it out and put it after the for loop.